packages/data-service/src/graphqlObservable.ts

```diff
@@ -0,0 +1,317 @@
+import "rxjs/add/operator/do";
+import "rxjs/add/operator/map";
+import "rxjs/add/operator/concatMap";
+
+import {
+  DefinitionNode,
+  DocumentNode,
+  getNamedType,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLSchema,
+  GraphQLType,
+  SelectionNode,
+  FieldNode,
+  GraphQLField,
+  GraphQLFieldResolver,
+  isTypeSystemDefinitionNode,
+  isTypeSystemExtensionNode,
+  Kind
+} from "graphql";
+
+// WARNING: This is NOT a spec complete graphql implementation
+// https://facebook.github.io/graphql/October2016/
+
+interface TypeMap {
+  [key: string]: GraphQLType;
+}
+
+interface Schema extends GraphQLSchema {
+  _typeMap: TypeMap;
+}
+
+interface OperationNode {
+  operation: "query" | "mutation";
+}
+
+type SchemaNode = SelectionNode | DefinitionNode;
+
+function isOperationDefinition(node: any): node is OperationNode {
+  return node.kind === Kind.OPERATION_DEFINITION;
+}
+function isFieldNode(node: SchemaNode): node is FieldNode {
+  return node.kind === Kind.FIELD;
+}
+
+// We don't treat OperationDefinitions as Definitions but as entry points for our execution
+function isDefinitionNode(node: SchemaNode): node is DefinitionNode {
+  return (
+    node.kind === Kind.FRAGMENT_DEFINITION ||
+    isTypeSystemDefinitionNode(node) ||
+    isTypeSystemExtensionNode(node)
+  );
+}
+
+interface FieldWithResolver extends GraphQLField<any, any> {
+  resolve: GraphQLFieldResolver<any, any, any>;
+}
+
+function isFieldWithResolver(
+  field: GraphQLField<any, any>
+): field is FieldWithResolver {
+  return field.resolve instanceof Function;
+}
+
+export function graphqlObservable<T = object>(
+  doc: DocumentNode,
+  schema: Schema,
+  context: object
+): Observable<{ data: T }> {
+  if (doc.definitions.length !== 1) {
+    return throwObservable("document root must have a single definition");
+  }
+  const types = schema._typeMap;
+
+  function resolve(
+    definition: SchemaNode,
+    context: object,
+    parent: any,
+    type: GraphQLType | null
+  ) {
+    if (isOperationDefinition(definition)) {
+      const nextType = getResultType(type, definition, parent);
+
+      return resolveResult(definition, context, null, nextType);
+    }
+
+    // The definition gives us the field to resolve
+    if (isFieldNode(definition)) {
+      const field = getField(type, definition);
+
+      // Something unexpcected was passed into getField
+      if (field === null) {
+        // TODO: find better worded error message
+        return throwObservable(
+          `field was not of the right type. Given type: ${type}`
+        );
+      }
+
+      const resolvedObservable = resolveField(
+        field,
+        definition,
+        context,
+        parent
+      );
+
+      if (!resolvedObservable) {
+        return throwObservable("resolver returned an empty value");
+      }
+
+      if (!(resolvedObservable instanceof Observable)) {
+        return throwObservable("resolver does not return an observable");
```

Comment:
I appreciate all the new assertions sprinkled in here. But these two checks (111, 115) are unreachable as written because resolveField always returns an Observable